libraries/AP_KDECAN/AP_KDECAN.cpp

```diff
@@ -53,603 +48,276 @@ const AP_Param::GroupInfo AP_KDECAN::var_info[] = {
+            break;
+    }
+}
 
-    if(!_can_iface->send(frame, AP_HAL::micros() + 1000000, 0)) {
-        debug_can(AP_CANManager::LOG_DEBUG, "couldn't send discovery message");
+void AP_KDECAN_Driver::update(const uint8_t num_poles)
+{
+    if (!hal.util->get_soft_armed() || _init.detected_bitmask == 0) {
         return;
     }
 
-    debug_can(AP_CANManager::LOG_DEBUG, "discovery message sent");
-
-    uint32_t start = AP_HAL::millis();
-
-    // wait 1 second for answers
-    while (AP_HAL::millis() - start < 1000) {
-        AP_HAL::CANFrame esc_id_frame {};
-        uint64_t rx_time;
-        AP_HAL::CANIface::CanIOFlags flags = 0;
-
-        int16_t n = _can_iface->receive(esc_id_frame, rx_time, flags);
-
-        if (n != 1) {
-            continue;
-        }
+#if HAL_WITH_ESC_TELEM
+    _telemetry.num_poles = num_poles;
+#endif
+    
+    uint16_t pwm[KDECAN_MAX_NUM_ESCS] {};
 
-        if (!esc_id_frame.isExtended()) {
+    uint8_t index = 0;
+    for (uint16_t i = 0; i < ARRAY_SIZE(pwm); i++) {
+        if ((_init.detected_bitmask & (1UL<<index)) == 0) {
             continue;
         }
-
-        if (esc_id_frame.dlc != 5) {
-            continue;
+        const SRV_Channel::Aux_servo_function_t function = SRV_Channels::channel_function(i);
+        if (SRV_Channel::is_motor(function)) {
+            pwm[index++] = SRV_Channels::srv_channel(i)->get_output_pwm();
         }
-
-        id.value = esc_id_frame.id & AP_HAL::CANFrame::MaskExtID;
-
-        if (id.source_id == BROADCAST_NODE_ID ||
-            id.source_id >= (KDECAN_MAX_NUM_ESCS + ESC_NODE_ID_FIRST) ||
-            id.destination_id != AUTOPILOT_NODE_ID ||
-            id.object_address != ESC_INFO_OBJ_ADDR) {
-            continue;
-        }
-
-        _esc_present_bitmask |= (1 << (id.source_id - ESC_NODE_ID_FIRST));
-        _esc_max_node_id = id.source_id - ESC_NODE_ID_FIRST + 1;
-
-        debug_can(AP_CANManager::LOG_DEBUG, "found ESC id %u", id.source_id);
     }
 
-    snprintf(_thread_name, sizeof(_thread_name), "kdecan_%u", driver_index);
-
-    // start thread for receiving and sending CAN frames
-    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_KDECAN::loop, void), _thread_name, 4096, AP_HAL::Scheduler::PRIORITY_CAN, 0)) {
-        debug_can(AP_CANManager::LOG_ERROR, "couldn't create thread");
-        return;
+    {
+        // queue the PWMs for loop()
+        WITH_SEMAPHORE(_output.sem);
+        memcpy(&_output.pwm, &pwm, sizeof(_output.pwm));
+        _output.is_new = true;
     }
 
-    _initialized = true;
-
-    debug_can(AP_CANManager::LOG_DEBUG, "init done");
-
-    return;
+#if KDECAN_USE_EVENTS
+    if (_output.thread_ctx != nullptr) {
+        // trigger the thread to wake up immediately
+        chEvtSignal(_output.thread_ctx, 1);
+    }
+#endif
 }
 
-void AP_KDECAN::loop()
+void AP_KDECAN_Driver::loop()
 {
-    uint64_t timeout;
+    hal.scheduler->delay(500);
```

Comment:
Why?

Reply:
Probably not be needed any more. I'll remove it.

Reply:
removed